packages/snaps-controllers/src/multichain/MultichainRouter.ts

```diff
@@ -0,0 +1,346 @@
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedMethods`,
+      (...args) => this.getSupportedMethods(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedAccounts`,
+      (...args) => this.getSupportedAccounts(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:isSupportedScope`,
+      (...args) => this.isSupportedScope(...args),
+    );
+  }
+
+  async #resolveRequestAddress(
+    snapId: SnapId,
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    try {
+      const result = (await this.#withSnapKeyring(async (keyring) =>
+        keyring.resolveAccountAddress(snapId, scope, request),
+      )) as { address: CaipAccountId } | null;
+      const address = result?.address;
+      return address ? parseCaipAccountId(address).address : null;
+    } catch {
+      throw rpcErrors.internal();
+    }
+  }
+
+  async #getAccountSnap(
+    connectedAddresses: CaipAccountId[],
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    const accounts = this.#messenger
+      .call('AccountsController:listMultichainAccounts', scope)
+      .filter(
+        (account: InternalAccount) =>
+          account.metadata.snap?.enabled &&
+          account.methods.includes(request.method),
+      ) as (InternalAccount & {
+      metadata: Required<InternalAccount['metadata']>;
+    })[];
+
+    // If no accounts can service the request, return null.
+    if (accounts.length === 0) {
+      return null;
+    }
+
+    const resolutionSnapId = accounts[0].metadata.snap.id;
+
+    // Attempt to resolve the address that should be used for signing.
+    const address = await this.#resolveRequestAddress(
+      resolutionSnapId,
+      scope,
+      request,
+    );
+
+    const parsedConnectedAddresses = connectedAddresses.map(
+      (connectedAddress) => parseCaipAccountId(connectedAddress).address,
+    );
+
+    // If we have a resolved address, try to find the selected account based on that
+    // otherwise, default to one of the connected accounts.
+    // TODO: Eventually let the user choose if we have more than one option for the account.
+    const selectedAccount = accounts.find(
+      (account) =>
+        parsedConnectedAddresses.includes(account.address) &&
+        (!address || account.address.toLowerCase() === address.toLowerCase()),
+    );
+
+    if (!selectedAccount) {
+      throw rpcErrors.invalidParams();
+    }
+
+    return {
+      accountId: selectedAccount.id,
+      snapId: selectedAccount.metadata.snap.id,
+    };
+  }
+
+  #getProtocolSnaps(scope: CaipChainId) {
+    const allSnaps = this.#messenger.call('SnapController:getAll');
+    const filteredSnaps = getRunnableSnaps(allSnaps);
+
+    return filteredSnaps.reduce<ProtocolSnap[]>((accumulator, snap) => {
+      const permissions = this.#messenger.call(
+        'PermissionController:getPermissions',
+        snap.id,
+      );
```

Comment:
Nit for readability:
```suggestion
      );

```